### PR TITLE
Shape preservation with ranges

### PR DESCRIPTION
**Pull request by vasslitvinov**
This continues the work in #9994 and #10758 for shapes that are ranges.

Given open questions about the design used in the above, see #11220,
the range case ultimately falls back to the currently-shapeless route.
As a result, while the shape is preserved, there is no parallelism.
On the upside, the design issue in #11220 is side-stepped by
retaining the existing behavior in that regard, i.e. the runtime type
for the resulting array's element type is taken from the first element.

If we wanted to move this case to the shape-ful implementation,
see a refactoring of the shapeful initCopy() in 151c047.
That refactoring ensures no memory leaks and no double-deletions.
Except the standin domain leaks as before.

Update the tests that depended on the old behavior.
Added more checking to the prototype partial-reduction implementation.
While there, updated a comment.

Testing:
- [x] linux64 -futures
- [x] gasnet - multilocale tests
- [x] valgrind
- [x] memleaks

### CODE REVIEW

**Comment by vasslitvinov** [comment]
@benharsh  would you review?

**Comment by benharsh** [comment]
Oops, forgot to mention: I think a new test should be added for those error messages if tests don't exist already.

**Comment by vasslitvinov** [comment]
Still not passing testing. Closing for now.
Will be back to it after the release.

**Comment by vasslitvinov** [comment]
Resuming this work.

**Comment by vasslitvinov** [comment]
@benharsh - I got new changes in 5f1feb1. How do they look to you?

>  I think a new test should be added for those error messages if tests don't exist already.

I think you were talking about the compilerError() additions [here](https://github.com/chapel-lang/chapel/pull/10942/files#diff-feddb5a31fea31a03458bbddb3164972l). How about adding tests for those when this code goes into production?

File: test/release/examples/benchmarks/shootout/fasta.chpl [hunk 1]
```diff
@@ -93,7 +93,7 @@ proc repeatMake(desc, alu, n) {
         s = [i in 0..(r+lineLength)] alu[i % r]: int(8);
 
   for i in 0..n by lineLength {
-    const lo = i % r + 1,
+    const lo = i % r,
```
Comment (benharsh) [inline test/release/examples/benchmarks/shootout/fasta.chpl]: Might want to let Brad know about this change so we can plan for the next re-submission.

**Pull request merged**